Migrate Profiles test from E2E to integrate test suite TP

dpham692 opened the pull request:
<!--
************ STOP!! ************
If this Pull Request is intended to fix a security vulnerability, DO NOT submit it! Instead, contact
the Apache Software Foundation Security Team at security@trafficcontrol.apache.org and follow the
guidelines at https://www.apache.org/security/ regarding vulnerability disclosure.
-->
## What does this PR (Pull Request) do?
<!-- Explain the changes you made here. If this fixes an Issue, identify it by
replacing the text in the checkbox item with the Issue number e.g.

- [x] This PR fixes #9001 OR is not related to any Issue

^ This will automatically close Issue number 9001 when the Pull Request is
merged (The '#' is important).

Be sure you check the box properly, see the "The following criteria are ALL
met by this PR" section for details.
-->

- [x] This PR fixes  is not related to any Issue <!-- You can check for an issue here: https://github.com/apache/trafficcontrol/issues -->


## Which Traffic Control components are affected by this PR?
<!-- Please delete all components from this list that are NOT affected by this
Pull Request. Also, feel free to add the name of a tool or script that is
affected but not on the list.

Additionally, if this Pull Request does NOT affect documentation, please
explain why documentation is not required. -->

- CDN in a Box
- Traffic Portal
- CI tests

## What is the best way to verify this PR?
<!-- Please include here ALL the steps necessary to test your Pull Request. If
it includes tests (and most should), outline here the steps needed to run the
tests. If not, lay out the manual testing procedure and please explain why
tests are unnecessary for this Pull Request. -->
Make sure all tests are migrate in integration/Data/profiles.ts directory.
https://github.com/apache/trafficcontrol/blob/master/traffic_portal/test/e2e_deprecated/profiles/profiles-spec.js
Go to traffic_portal/test/integration and run
`npm test -- --params.baseUrl https://localhost --capabilities.shardTestFiles true` and make sure test all pass.
Make sure CI tests pass also.
## If this is a bug fix, what versions of Traffic Control are affected?
<!-- If this PR fixes a bug, please list here all of the affected versions - to
the best of your knowledge. It's also pretty helpful to include a commit hash
of where 'master' is at the time this PR is opened (if it affects master),
because what 'master' means will change over time. For example, if this PR
fixes a bug that's present in master (at commit hash '1df853c8'), in v4.0.0,
and in the current 4.0.1 Release candidate (e.g. RC1), then this list would
look like:

- master (1df853c8)
- 4.0.0
- 4.0.1 (RC1)

If you don't know what other versions might have this bug, AND don't know how
to find the commit hash of 'master', then feel free to leave this section
blank (or, preferably, delete it entirely).
 -->


## The following criteria are ALL met by this PR
<!-- Check the boxes to signify that the associated statement is true. To
"check a box", replace the space inside of the square brackets with an 'x'.
e.g.

- [ x] <- Wrong
- [x ] <- Wrong
- [] <- Wrong
- [*] <- Wrong
- [x] <- Correct!

-->

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [ ] This PR includes documentation OR I have explained why documentation is unnecessary
- [ ] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [ ] This PR includes any and all required license headers
- [ ] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)


## Additional Information
<!-- If you would like to include any additional information on the PR for
potential reviewers please put it here.

Some examples of this would be:

- Before and after screenshots/gifs of the Traffic Portal if it is affected
- Links to other dependent Pull Requests
- References to relevant context (e.g. new/updates to dependent libraries,
mailing list records, blueprints)

Feel free to leave this section blank (or, preferably, delete it entirely).
-->

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

ocket8888 commented on traffic_portal/test/integration/specs/Profiles.spec.ts:
```diff
@@ -46,6 +46,24 @@ profiles.tests.forEach(async profilesData => {
+                        await profilesPage.OpenProfilesPage();
+                    }else{
+                        expect(await profilesPage.ToggleTableColumn(toggle.Name)).toBe(true);
+                        await profilesPage.OpenProfilesPage();
```
why are we re-opening the Profiles page at the end of all of these tests? These don't appear to be doing navigation, but either way a better way to reset the location after every test is with an `afterEach` call.

dpham692 replied:
I do the reopening Profiles page after each test because if the tests fail it will go back to the Profiles Page and continue with other tests.

dpham692 commented:
@ocket8888 I added aftereach and afterall like you said. Change it multiple way to fit the test suit because it has to log in and log out few times. I also tested it a couple times and make sure the spec clean.

ocket8888 commented on traffic_portal/test/integration/specs/Profiles.spec.ts:
```diff
@@ -33,47 +33,57 @@ describe('Setup API for Profiles', () => {
-            profilesData.add.forEach(add => {
-                it(add.description, async () => {
-                    expect(await profilesPage.CreateProfile(add)).toBeTruthy();
+    profiles.tests.forEach(async profilesData => {
+        profilesData.logins.forEach(login => {
```
The restructuring here confuses me - this used to be a separate `describe` for each login for each profile, but now it's all under the setup `describe`? It looks that way, but then the `describe`'s description isn't accurate, it's not just setting up it's actually doing all the Profile testing. That's not a terrible way to do it, but given that these specs are asynchronous, there's no guarantee that any one of them will run before any other (and in general you shouldn't assume that to allow for greater degrees of parallelism); that is to say, you don't know that the setup is run to completion before the other tests run. So it should probably go in a `beforeAll` - I actually think it ought to have been done that way in the first place, but the subsequent suite method was working so that change was unnecessary until the `it`s were sharing a `describe`.

By the same reasoning, the teardown logic in the "cleanup" spec should also probably be in an `afterAll`.

Further, we're re-adding `afterEach` `describe`, but it can actually just be added once in the outer `describe` since all spec blocks in a nested suite are also in the outer suite, and so are affected by `beforeEach` and `afterEach` in that outer block.

dpham692 replied:
I will fix the structure back to 3 describes. The **beforeAll** is a good idea, I will add that in.  I have to test this out for the **afterall** to answer you because I read somewhere that you can only have one **afterall** and I already use that for the log out function (log out 3 times for each users). @ocket8888

ocket8888 replied:
You can have one `afterAll` per `describe`, I think - but also you can just do multiple things in one `afterAll` if that's what you're talking about.

ocket8888 commented on traffic_portal/test/integration/specs/Profiles.spec.ts:
```diff
@@ -29,49 +29,70 @@ const loginPage = new LoginPage();
-        await api.UseAPI(profiles.setup);
-    });
-});
+beforeAll(async function () {
+    await api.UseAPI(profiles.setup);
+})
+
```
So you don't wanna do this, because putting a `beforeAll` in the global scope like that is going to run this before every single `it` in the entire integration testing suite.

dpham692 replied:
I see that's why the test took so long. I think I will add a 1 beforeAll in the config before all the test run.

ocket8888 replied:
Could be, hard to say.

What would that `beforeAll` in the config module be doing? It shouldn't navigate to the Profiles page, that'll have the same problem.

dpham692 commented:
@ocket8888 I changed it back to 3 describe, fix **afterall** and **aftereach**. For the **beforeall** I removed it out right now and will figure how to do with it in a future.
